test/__tests__/patterns.js

```diff
@@ -0,0 +1,72 @@
+import { runCli } from "../utils";
+
+// MIGRATION: Worked as is.
+describe("multiple patterns", () => {
+  runCli("patterns", ["directory/**/*.js", "other-directory/**/*.js", "-l"]).test({
+    status: 1,
+  });
+});
+
+// MIGRATION: Error is no longer reported when there are other matches, even with
+// `--error-on-unmatched-pattern`. Adjusted the stderr snapshot for now.
+// Should we change the logic to report the invalid patterns regardless?
+// Changed tha status to 1.
```

Comment:
Looking at the code in the rewrite we report an error only if nothing is actually matched, not if something is matched but it's subsequently ignored.

There may be a difference here depending on how many patterns we are trying to match, like does the old CLI throw if any pattern is unmatched, or if every pattern in the list of patterns is unmatched? Worth checking which one it is before deciding what to do here.

Reply:
This test case in the old CLI returns 2, the matches for `directory/**/*.js` in stdout, and `[error] No files matching the pattern were found: "non-existent.js".` in stderr. The new version is kinda counter-intuitive, as it returns 1, but doesn't write anything to stderr, because we got some matches from the valid pattern. The code responsible for this is on lines 200-204 in `index.ts`.

Reply:
So we do want to return an error when a single pattern isn't finding any _non ignored_ fils within the list of patterns, right? 🤔 Or is whether the found files are ignored or not not a factor in the old CLI/

Reply:
Let's consider the case where there are two patterns, one that generates some matches, and that doesn't. In my opinion, the latter isn't indicative of a failure – there simply aren't files that match it, and that's fine. So I think the best course of action would be to return 0 (indicating success), write the matches to stdout, and write nothing to stderr.

In this paradigm (not sure whether it's the right word here), even having a single pattern that doesn't match anything wouldn't be a failure – it would simply write nothing both to stdout and stderr, and return 0. Only if we set `--error-on-unmatched-pattern`, we should write to stderr in both cases, also including which pattern didn't match in the message.

What do you think?

Reply:
I don't have as much context as I had when I wrote this code, but it looks like erroring on unmatched patters is the default behavior in v3 and v4.

I think what you are proposing makes sense to me, but since this aims to be a largely backwards compatible rewrite (though not fully) I think if we can align with v3 it seems better for our goals with this 🤔

I don't remember what the difference is exactly here, we only error if the sum of all globs didn't find any matches, while v3 errors if any globs didn't find any matches? If so that's a bit problematic to handle as we would probably want some extra metadata for this from tiny-readdir-glob, to avoid searching with each glob individually and then merging their results, but it should be doable.

Reply:
Maybe we should add a comment here and investigate v3 vs v4 (and decide whether to add extra metadata to `tiny-readdir-glob`) in a separate PR so we could land this one? From when I first worked on this, I seem to remember there were some problems with `--error-on-unmatched-pattern` too, so it would make sense to research within a smaller scope.